boundedint/int256_test.go

```diff
@@ -0,0 +1,173 @@
+		expectsError: false,
+	},
+	// negative numbers
+	{
+		name:         "base -1",
+		baseInteger:  big.NewInt(-1),
+		expectsError: false,
+	},
+	{
+		name:         "base -1 * 2^8",
+		baseInteger:  new(big.Int).Mul(big.NewInt(-1), new(big.Int).Exp(big.NewInt(2), big.NewInt(8), nil)),
+		expectsError: false,
+	},
+	{
+		name:         "base -1 * 2^64",
+		baseInteger:  new(big.Int).Mul(big.NewInt(-1), new(big.Int).Exp(big.NewInt(2), big.NewInt(64), nil)),
+		expectsError: false,
+	},
+	{
+		name:         "base -1 * 2^255",
+		baseInteger:  new(big.Int).Mul(big.NewInt(-1), new(big.Int).Exp(big.NewInt(2), big.NewInt(255), nil)),
+		expectsError: false,
+	},
+	{
+		name:         "base -1 * 2^255 - 1",
+		baseInteger:  new(big.Int).Sub(new(big.Int).Mul(big.NewInt(-1), new(big.Int).Exp(big.NewInt(2), big.NewInt(255), nil)), big.NewInt(1)),
+		expectsError: true,
+	},
+	{
+		name:         "base -1 * 2^512",
+		baseInteger:  new(big.Int).Mul(big.NewInt(-1), new(big.Int).Exp(big.NewInt(2), big.NewInt(512), nil)),
+		expectsError: true,
+	},
+	// positive numbers
+	{
+		name:         "base 1",
+		baseInteger:  big.NewInt(1),
+		expectsError: false,
+	},
+	{
+		name:         "base 2^8",
+		baseInteger:  new(big.Int).Exp(big.NewInt(2), big.NewInt(8), nil),
+		expectsError: false,
+	},
+	{
+		name:         "base 2^128",
+		baseInteger:  new(big.Int).Exp(big.NewInt(2), big.NewInt(128), nil),
+		expectsError: false,
+	},
+	{
+		name:         "base 2^255 - 1",
+		baseInteger:  new(big.Int).Sub(new(big.Int).Exp(big.NewInt(2), big.NewInt(255), nil), big.NewInt(1)),
+		expectsError: false,
+	},
+	{
+		name:         "base 2^255",
+		baseInteger:  new(big.Int).Exp(big.NewInt(2), big.NewInt(255), nil),
+		expectsError: true,
+	},
+	{
+		name:         "base 2^512",
+		baseInteger:  new(big.Int).Exp(big.NewInt(2), big.NewInt(512), nil),
+		expectsError: true,
+	},
+}
+
+// TestSet tests the creation of valid and invalid Int256 structs by calling the Set function
+func TestInt256Set(t *testing.T) {
+	for _, tc := range int256TestCases {
+		t.Run(tc.name, func(t *testing.T) {
+			result, err := NewInt256().Set(*tc.baseInteger)
+			if tc.expectsError && err == nil {
+				t.Fatalf("expected error when creating new Int256, but got none")
+			}
+			if !tc.expectsError {
+				if err != nil {
+					t.Fatalf("got unexpected error when creating new Int256: %v", err)
+				}
+				resultValue := result.Value()
+				if (&resultValue).Cmp(tc.baseInteger) != 0 {
+					t.Fatalf("expected value of %v, got %v instead", tc.baseInteger, result.value)
+				}
+			}
+		})
+	}
+}
+
+// TestCopy tests the duplication of an existing Int256 variable
+func TestInt256Copy(t *testing.T) {
+	for _, tc := range int256TestCases {
+		t.Run(tc.name, func(t *testing.T) {
+			if !tc.expectsError {
+				r, err := NewInt256().Set(*tc.baseInteger)
+				if err != nil {
+					t.Fatalf("got unexpected error when creating new Int256: %v", err)
+				}
+				c := NewInt256().Copy(r)
+
+				if !c.Equals(r) {
```

Comment:
although this tests whether an object has been copied (a term whose semantics are already problematic), it does _not_ test that this copy is not in fact just a pointer to the old value. so after copying and comparing, i would also adjust the value of the variable that was copied _from_ then check that they _do not equal_. also, you don't need to run all cases again - a single unit test would do.

Reply:
fair enough.

i've changed the test to use just 1 arbitrary value, and also to check if the copy changes when the original variable changes—in which case a failure occurs.

thanks for the input